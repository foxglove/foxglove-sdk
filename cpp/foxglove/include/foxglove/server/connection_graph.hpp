#pragma once

#include <foxglove/error.hpp>

#include <cstdint>
#include <memory>
#include <string>

struct foxglove_connection_graph;

namespace foxglove {

class ConnectionGraph final {
  friend class WebSocketServer;

public:
  ConnectionGraph();

  FoxgloveError setPublishedTopic(
    std::string_view topic, const std::vector<std::string>& publisher_ids
  );
  FoxgloveError setSubscribedTopic(
    std::string_view topic, const std::vector<std::string>& subscriber_ids
  );
  FoxgloveError setAdvertisedService(
    std::string_view service, const std::vector<std::string>& provider_ids
  );

private:
  std::unique_ptr<foxglove_connection_graph, void (*)(foxglove_connection_graph*)> impl_;
};

}  // namespace foxglove
