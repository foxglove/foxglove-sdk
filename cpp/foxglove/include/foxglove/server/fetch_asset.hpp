#pragma once

#include <functional>
#include <memory>
#include <string_view>
#include <vector>

struct foxglove_fetch_asset_responder;

namespace foxglove {

/**
 * A fetch asset responder.
 *
 * This is the means by which a fetch asset implementation responds to a request
 * from a client. Each request is paired with a unique responder instance, and
 * must be used exactly once.
 */
class FetchAssetResponder final {
public:
  /**
   * Sends asset data to the client.
   */
  void respondOk(const std::vector<std::byte>& data) && noexcept;

  /**
   * Sends an error message to the client.
   */
  void respondError(std::string_view message) && noexcept;

  // Default destructor & move, disable copy.
  ~FetchAssetResponder() = default;
  FetchAssetResponder(FetchAssetResponder&&) noexcept = default;
  FetchAssetResponder& operator=(FetchAssetResponder&&) noexcept = default;
  FetchAssetResponder(const FetchAssetResponder&) = delete;
  FetchAssetResponder& operator=(const FetchAssetResponder&) = delete;

private:
  friend class WebSocketServer;

  explicit FetchAssetResponder(foxglove_fetch_asset_responder* ptr)
      : impl_(ptr) {}

  struct Deleter {
    void operator()(foxglove_fetch_asset_responder* ptr) const noexcept;
  };
  std::unique_ptr<foxglove_fetch_asset_responder, Deleter> impl_;
};

/**
 * A fetch asset handler callback.
 *
 * This callback is invoked from the client's main poll loop and must not block.
 * If blocking or long-running behavior is required, the implementation should
 * return immediately and handle the request asynchronously.
 *
 * The `responder` represents an unfulfilled response. The implementation must
 * eventually call either `respondOk` or `respondError`, exactly once, in order
 * to complete the request. It is safe to invoke these completion methods
 * synchronously from the context of the callback.
 */
using FetchAssetHandler =
  std::function<void(std::string_view uri, FetchAssetResponder&& responder)>;

}  // namespace foxglove
