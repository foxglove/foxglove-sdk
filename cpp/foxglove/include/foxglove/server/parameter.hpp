#pragma once

#include <foxglove/error.hpp>

#include <cstdint>
#include <map>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <variant>
#include <vector>

struct foxglove_parameter_value;
struct foxglove_parameter;
struct foxglove_parameter_array;

namespace foxglove {

/**
 * A parameter type.
 *
 * This enum is used to disambiguate `Parameter` values, in situations where the
 * wire representation is ambiguous.
 */
enum class ParameterType : uint8_t {
  // The parameter value can be inferred from the inner parameter value tag.
  None,
  // An array of bytes.
  ByteArray,
  // A decimal or integer value that can be represented as a `float64`.
  Float64,
  // An array of decimal or integer values that can be represented as `float64`s.
  Float64Array,
};

/**
 * A view over an unowned parameter value.
 *
 * This lifetime of this view is tied to the `ParameterValue` from which it was
 * derived. It is the caller's responsibility to ensure the validity of this
 * lifetime when accessing the view.
 */
class ParameterValueView final {
public:
  using Array = std::vector<ParameterValueView>;
  using Dict = std::map<std::string, ParameterValueView>;
  using Value = std::variant<double, bool, std::string, Array, Dict>;

  // Creates a deep clone of this parameter value.
  [[nodiscard]] class ParameterValue clone() const;

  // Returns a variant representation of the value.
  [[nodiscard]] Value value() const;

  // Value type checker.
  template<typename T>
  [[nodiscard]] bool is() const {
    auto value = this->value();
    return std::holds_alternative<T>(value);
  }

  // Value extractor.
  template<typename T>
  [[nodiscard]] T get() const {
    throw std::runtime_error("Unsupported type");
  }
  template<>
  [[nodiscard]] ParameterValueView get<ParameterValueView>() const {
    return *this;
  }
  template<>
  [[nodiscard]] double get<double>() const {
    return std::get<double>(this->value());
  }
  template<>
  [[nodiscard]] bool get<bool>() const {
    return std::get<bool>(this->value());
  }
  template<>
  [[nodiscard]] std::string get<std::string>() const {
    return std::get<std::string>(this->value());
  }
  template<>
  [[nodiscard]] Array get<Array>() const {
    return std::get<Array>(this->value());
  }
  template<>
  [[nodiscard]] Dict get<Dict>() const {
    return std::get<Dict>(this->value());
  }

private:
  friend class ParameterView;
  friend class ParameterValue;

  const foxglove_parameter_value* _impl;
  explicit ParameterValueView(const foxglove_parameter_value* ptr)
      : _impl(ptr) {}
};

/**
 * An owned parameter value.
 */
class ParameterValue final {
public:
  // Constructors
  explicit ParameterValue(double value);
  explicit ParameterValue(bool value);
  explicit ParameterValue(std::string_view value);
  explicit ParameterValue(const char* value)
      : ParameterValue(std::string_view(value)) {}
  explicit ParameterValue(std::vector<ParameterValue> values);
  explicit ParameterValue(std::map<std::string, ParameterValue> values);

  // Default destructor & move, disable copy.
  ~ParameterValue() = default;
  ParameterValue(ParameterValue&& other) noexcept = default;
  ParameterValue& operator=(ParameterValue&& other) noexcept = default;
  ParameterValue(const ParameterValue&) = delete;
  ParameterValue& operator=(const ParameterValue&) = delete;

  // Creates a deep clone of this parameter value.
  [[nodiscard]] ParameterValue clone() const {
    return this->view().clone();
  }

  // Accessors
  [[nodiscard]] ParameterValueView view() const noexcept;
  [[nodiscard]] ParameterValueView::Value value() const {
    return this->view().value();
  }

  // Value type checker.
  template<typename T>
  [[nodiscard]] bool is() const {
    return this->view().is<T>();
  }

  // Value extractor.
  template<typename T>
  [[nodiscard]] T get() const {
    return this->view().get<T>();
  }

private:
  friend class ParameterValueView;
  friend class Parameter;

  struct Deleter {
    void operator()(foxglove_parameter_value* ptr) const noexcept;
  };

  std::unique_ptr<foxglove_parameter_value, Deleter> _impl;

  // Constructor from raw pointer.
  explicit ParameterValue(foxglove_parameter_value*);

  // Releases ownership of the underlying storage.
  [[nodiscard]] foxglove_parameter_value* release() noexcept;
};

/**
 * A view over an unowned parameter.
 *
 * This lifetime of this view is tied to the `Parameter` from which it was
 * derived. It is the caller's responsibility to ensure the validity of this
 * lifetime when accessing the view.
 */
class ParameterView final {
public:
  // Creates a deep clone of this parameter.
  [[nodiscard]] class Parameter clone() const;

  // Accessors
  [[nodiscard]] std::string_view name() const noexcept;
  [[nodiscard]] ParameterType type() const noexcept;
  [[nodiscard]] std::optional<ParameterValueView> value() const noexcept;
  [[nodiscard]] bool hasValue() const noexcept {
    return this->value().has_value();
  };

  // Value type checkers.
  template<typename T>
  [[nodiscard]] bool is() const {
    return this->hasValue() && this->value()->is<T>();
  }
  template<>
  [[nodiscard]] bool is<std::string>() const {
    return this->hasValue() && this->type() != ParameterType::ByteArray &&
           this->value()->is<std::string>();
  }
  [[nodiscard]] bool isArray() const noexcept {
    return this->hasValue() && this->value()->is<ParameterValueView::Array>();
  }
  [[nodiscard]] bool isDict() const noexcept {
    return this->hasValue() && this->value()->is<ParameterValueView::Dict>();
  }
  [[nodiscard]] bool isFloat64Array() const noexcept {
    return this->hasValue() && this->type() == ParameterType::Float64Array && this->isArray();
  }
  [[nodiscard]] bool isByteArray() const noexcept {
    return this->hasValue() && this->type() == ParameterType::ByteArray &&
           this->value()->is<std::string>();
  }

  // Value extractor.
  template<typename T>
  [[nodiscard]] T get() const {
    auto value = this->value();
    if (!value) {
      throw std::bad_optional_access();
    }
    return value->template get<T>();
  }

  // Value array extractor.
  template<typename T>
  [[nodiscard]] std::vector<T> getArray() const {
    auto value = this->value();
    if (!value) {
      throw std::bad_optional_access();
    }
    const auto& arr = value->get<ParameterValueView::Array>();
    std::vector<T> result;
    result.reserve(arr.size());
    for (const auto& elem : arr) {
      result.push_back(elem.get<T>());
    }
    return result;
  }

  // Value dict extractor.
  template<typename T>
  [[nodiscard]] std::map<std::string, T> getDict() const {
    auto value = this->value();
    if (!value) {
      throw std::bad_optional_access();
    }
    const auto& dict = value->get<ParameterValueView::Dict>();
    std::map<std::string, T> result;
    for (const auto& elem : dict) {
      std::string key(elem.first);
      auto value = elem.second.get<T>();
      result.emplace(key, value);
    }
    return result;
  }

  // Value byte array extractor.
  [[nodiscard]] FoxgloveResult<std::vector<std::byte>> getByteArray() const;

private:
  friend class Parameter;
  friend class ParameterArrayView;

  const foxglove_parameter* _impl;
  explicit ParameterView(const foxglove_parameter* ptr)
      : _impl(ptr) {}
};

/**
 * An owned parameter.
 */
class Parameter final {
public:
  explicit Parameter(std::string_view);
  explicit Parameter(std::string_view name, double value);
  explicit Parameter(std::string_view name, bool value);
  explicit Parameter(std::string_view name, std::string_view value);
  explicit Parameter(std::string_view name, const char* value)
      : Parameter(name, std::string_view(value)) {}
  explicit Parameter(std::string_view name, const uint8_t* data, size_t data_length);
  explicit Parameter(std::string_view name, const std::vector<std::byte>& bytes)
      : Parameter(name, reinterpret_cast<const uint8_t*>(bytes.data()), bytes.size()) {}
  explicit Parameter(std::string_view name, const std::vector<double>& values);
  explicit Parameter(std::string_view name, std::map<std::string, ParameterValue> values);
  explicit Parameter(std::string_view name, ParameterType, ParameterValue&& value);

  // Default destructor & move, disable copy.
  ~Parameter() = default;
  Parameter(Parameter&& other) noexcept = default;
  Parameter& operator=(Parameter&& other) noexcept = default;
  Parameter(const Parameter&) = delete;
  Parameter& operator=(const Parameter&) = delete;

  // Creates a deep clone of this parameter.
  [[nodiscard]] Parameter clone() const {
    return this->view().clone();
  }

  // Accessors
  [[nodiscard]] ParameterView view() const noexcept;
  [[nodiscard]] std::string_view name() const noexcept {
    return this->view().name();
  }
  [[nodiscard]] ParameterType type() const noexcept {
    return this->view().type();
  }
  [[nodiscard]] std::optional<ParameterValueView> value() const noexcept {
    return this->view().value();
  }
  [[nodiscard]] bool hasValue() const noexcept {
    return this->view().hasValue();
  };

  // Value type checkers.
  template<typename T>
  [[nodiscard]] bool is() const {
    return this->view().is<T>();
  }
  [[nodiscard]] bool isArray() const noexcept {
    return this->view().isArray();
  }
  [[nodiscard]] bool isDict() const noexcept {
    return this->view().isDict();
  }
  [[nodiscard]] bool isFloat64Array() const noexcept {
    return this->view().isFloat64Array();
  }
  [[nodiscard]] bool isByteArray() const noexcept {
    return this->view().isByteArray();
  }

  // Value extractor.
  template<typename T>
  [[nodiscard]] T get() const {
    return this->view().get<T>();
  }

  // Value array extractor.
  template<typename T>
  [[nodiscard]] std::vector<T> getArray() const {
    return this->view().getArray<T>();
  }

  // Value dict extractor.
  template<typename T>
  [[nodiscard]] std::map<std::string, T> getDict() const {
    return this->view().getDict<T>();
  }

  // Value byte array extractor.
  [[nodiscard]] FoxgloveResult<std::vector<std::byte>> getByteArray() const {
    return this->view().getByteArray();
  }

private:
  friend class ParameterView;
  friend class ParameterArray;

  struct Deleter {
    void operator()(foxglove_parameter* ptr) const noexcept;
  };

  std::unique_ptr<foxglove_parameter, Deleter> _impl;

  explicit Parameter(foxglove_parameter* param);

  // Releases ownership of the underlying storage.
  [[nodiscard]] foxglove_parameter* release() noexcept;
};

/**
 * A view over an unowned parameter array.
 *
 * This lifetime of this view is tied to the `ParameterArray` from which it was
 * derived. It is the caller's responsibility to ensure the validity of this
 * lifetime when accessing the view.
 */
class ParameterArrayView final {
public:
  explicit ParameterArrayView(const foxglove_parameter_array*);

  [[nodiscard]] std::vector<ParameterView> parameters() const;

private:
  const foxglove_parameter_array* _impl;
};

/**
 * An owned parameter array.
 */
class ParameterArray final {
public:
  explicit ParameterArray(std::vector<Parameter>&&);

  // Default destructor & move, disable copy.
  ~ParameterArray() = default;
  ParameterArray(ParameterArray&& other) noexcept = default;
  ParameterArray& operator=(ParameterArray&& other) noexcept = default;
  ParameterArray(const ParameterArray&) = delete;
  ParameterArray& operator=(const ParameterArray&) = delete;

  // Accessors
  [[nodiscard]] ParameterArrayView view() const noexcept;
  [[nodiscard]] std::vector<ParameterView> parameters() const {
    return this->view().parameters();
  }

  // Releases ownership of the underlying storage.
  [[nodiscard]] foxglove_parameter_array* release() noexcept;

private:
  struct Deleter {
    void operator()(foxglove_parameter_array* ptr) const noexcept;
  };

  std::unique_ptr<foxglove_parameter_array, Deleter> _impl;
};

}  // namespace foxglove
