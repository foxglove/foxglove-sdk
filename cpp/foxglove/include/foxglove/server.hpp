#pragma once

#include <foxglove/context.hpp>
#include <foxglove/error.hpp>
#include <foxglove/server/connection_graph.hpp>
#include <foxglove/server/fetch_asset.hpp>
#include <foxglove/server/parameter.hpp>
#include <foxglove/server/service.hpp>

#include <cstdint>
#include <functional>
#include <memory>
#include <string>

enum foxglove_error : uint8_t;
struct foxglove_websocket_server;
struct foxglove_connection_graph;

namespace foxglove {

/// @brief A channel advertised by a client.
struct ClientChannel {
  /// @brief The ID of the channel.
  uint32_t id;
  /// @brief The topic of the channel.
  std::string_view topic;
  /// @brief The encoding of the channel.
  std::string_view encoding;
  /// @brief The name of the schema of the channel.
  std::string_view schema_name;
  /// @brief The encoding of the schema of the channel.
  std::string_view schema_encoding;
  /// @brief The schema of the channel.
  const std::byte* schema;
  /// @brief The length of the schema of the channel.
  size_t schema_len;
};

/// @brief The capabilities of a WebSocket server.
///
/// A server may advertise certain capabilities to clients and provide related functionality
/// in WebSocketServerCallbacks.
enum class WebSocketServerCapabilities : uint8_t {
  /// Allow clients to advertise channels to send data messages to the server.
  ClientPublish = 1 << 0,
  /// Allow clients to subscribe and make connection graph updates
  ConnectionGraph = 1 << 1,
  /// Allow clients to get & set parameters.
  Parameters = 1 << 2,
  /// Inform clients about the latest server time.
  ///
  /// This allows accelerated, slowed, or stepped control over the progress of time. If the
  /// server publishes time data, then timestamps of published messages must originate from the
  /// same time source.
  Time = 1 << 3,
  /// Allow clients to call services.
  Services = 1 << 4,
  /// Allow clients to request assets. If you supply an asset handler to the
  /// server, this capability will be advertised automatically.
  Assets = 1 << 5,
};

/// @brief Combine two capabilities.
inline WebSocketServerCapabilities operator|(
  WebSocketServerCapabilities a, WebSocketServerCapabilities b
) {
  return WebSocketServerCapabilities(uint8_t(a) | uint8_t(b));
}

/// @brief Check if a capability is set.
inline WebSocketServerCapabilities operator&(
  WebSocketServerCapabilities a, WebSocketServerCapabilities b
) {
  return WebSocketServerCapabilities(uint8_t(a) & uint8_t(b));
}

/// @brief The callback interface for a WebSocket server.
///
/// These methods are invoked from the client's main poll loop and must be as low-latency as
/// possible.
struct WebSocketServerCallbacks {
  /// @brief Callback invoked when a client subscribes to a channel.
  ///
  /// Only invoked if the channel is associated with the server and isn't already subscribed to by
  /// the client.
  std::function<void(uint64_t channel_id)> onSubscribe;

  /// @brief Callback invoked when a client unsubscribes from a channel.
  ///
  /// Only invoked for channels that had an active subscription from the client.
  std::function<void(uint64_t channel_id)> onUnsubscribe;

  /// @brief Callback invoked when a client advertises a client channel.
  ///
  /// Requires the capability WebSocketServerCapabilities::ClientPublish
  std::function<void(uint32_t client_id, const ClientChannel& channel)> onClientAdvertise;

  /// @brief Callback invoked when a client message is received
  std::function<
    void(uint32_t client_id, uint32_t client_channel_id, const std::byte* data, size_t data_len)>
    onMessageData;

  /// @brief Callback invoked when a client unadvertises a client channel.
  ///
  /// Requires the capability WebSocketServerCapabilities::ClientPublish
  std::function<void(uint32_t client_id, uint32_t client_channel_id)> onClientUnadvertise;

  /// @brief Callback invoked when a client requests parameters.
  ///
  /// Requires the capability WebSocketServerCapabilities::Parameters.
  ///
  /// @param client_id The client ID.
  /// @param request_id A request ID unique to this client. May be NULL.
  /// @param param_names A list of parameter names to fetch. If empty, this
  /// method should return all parameters.
  std::function<std::vector<Parameter>(
    uint32_t client_id, std::optional<std::string_view> request_id,
    const std::vector<std::string_view>& param_names
  )>
    onGetParameters;

  /// @brief Callback invoked when a client sets parameters.
  ///
  /// Requires the capability WebSocketServerCapabilities::Parameters.
  ///
  /// This function should return the updated parameters. All clients subscribed
  /// to updates for the returned parameters will be notified.
  ///
  /// @param client_id The client ID.
  /// @param request_id A request ID unique to this client. May be NULL.
  /// @param param_names A list of updated parameter values.
  std::function<std::vector<Parameter>(
    uint32_t client_id, std::optional<std::string_view> request_id,
    const std::vector<ParameterView>& params
  )>
    onSetParameters;

  /// @brief Callback invoked when a client subscribes to the named parameters
  /// for the first time.
  ///
  /// Requires the capability WebSocketServerCapabilities::Parameters.
  ///
  /// @param param_names A list of parameter names.
  std::function<void(const std::vector<std::string_view>& param_names)> onParametersSubscribe;

  /// @brief Callback invoked when the last client unsubscribes from the named
  /// parameters.
  ///
  /// Requires the capability WebSocketServerCapabilities::Parameters.
  ///
  /// @param param_names A list of parameter names.
  std::function<void(const std::vector<std::string_view>& param_names)> onParametersUnsubscribe;

  /// @brief Callback invoked when a client requests connection graph updates.
  ///
  /// Requires the capability WebSocketServerCapabilities::ConnectionGraph
  std::function<void()> onConnectionGraphSubscribe;

  /// @brief Callback invoked when a client unsubscribes from connection graph updates.
  ///
  /// Requires the capability WebSocketServerCapabilities::ConnectionGraph
  std::function<void()> onConnectionGraphUnsubscribe;
};

/// @brief Options for a WebSocket server.
struct WebSocketServerOptions {
  friend class WebSocketServer;

  /// @brief The logging context for this server.
  Context context;
  /// @brief The name of the server.
  std::string name;
  /// @brief The host address of the server.
  std::string host = "127.0.0.1";
  /// @brief The port of the server. Default is 8765, which matches the Foxglove app.
  uint16_t port = 8765;
  /// @brief The callbacks of the server.
  WebSocketServerCallbacks callbacks;
  /// @brief The capabilities of the server.
  WebSocketServerCapabilities capabilities = WebSocketServerCapabilities(0);
  /// @brief The supported encodings of the server.
  std::vector<std::string> supported_encodings;
  /// @brief A fetch asset handler callback.
  FetchAssetHandler fetch_asset;
};

/// @brief A WebSocket server for visualization in Foxglove.
///
/// After your server is started, you can open the Foxglove app to visualize your data. See
/// [Connecting to data].
///
/// [Connecting to data]: https://docs.foxglove.dev/docs/connecting-to-data/introduction
class WebSocketServer final {
public:
  /// @brief Create a new WebSocket server with the given options.
  static FoxgloveResult<WebSocketServer> create(WebSocketServerOptions&& options);

  /// Get the port on which the server is listening.
  [[nodiscard]] uint16_t port() const;

  /// @brief Gracefully shut down the websocket server.
  FoxgloveError stop();

  /// @brief Publishes the current server timestamp to all clients.
  ///
  /// Requires the capability WebSocketServerCapabilities::Time.
  ///
  /// @param timestamp_nanos An epoch offset in nanoseconds.
  void broadcastTime(uint64_t timestamp_nanos) const noexcept;

  /// @brief Advertises support for the provided service.
  ///
  /// This service will be available for clients to use until it is removed with
  /// `removeService()`, or the server is stopped.
  ///
  /// This method will fail for various reasons, with the following error codes:
  ///
  /// - `DuplicateService`: A service with the same name is already registered.
  /// - `MissingRequestedEncoding`: The service didn't declare a request
  ///   encoding, and the server was not configured with a global list of
  ///   supported encodings.
  /// - `ServicesNotSupported`: The server was not convfigured with the
  ///   `Services` capability.
  ///
  /// @param service The service to add.
  [[nodiscard]] FoxgloveError addService(Service&& service) const noexcept;

  /// @brief Removes a service that was previously advertised.
  ///
  /// This method will fail with `FoxgloveError::Utf8Error` if the service name
  /// is not a valid UTF-8 string.
  ///
  /// @param name The name of the service to remove.
  [[nodiscard]] FoxgloveError removeService(std::string_view name) const noexcept;

  /// @brief Publishes parameter values to all subscribed clients.
  ///
  /// Requires the capability WebSocketServerCapabilities::Parameters.
  ///
  /// @param params Updated parameters.
  void publishParameterValues(std::vector<Parameter>&& params);

  /// @brief Publish a connection graph to all subscribed clients.
  ///
  /// @param graph The connection graph to publish.
  ///
  /// This requires the capability WebSocketServerCapabilities::ConnectionGraph
  void publishConnectionGraph(ConnectionGraph& graph);

private:
  WebSocketServer(
    foxglove_websocket_server* server, std::unique_ptr<WebSocketServerCallbacks> callbacks,
    std::unique_ptr<FetchAssetHandler> fetch_asset
  );

  std::unique_ptr<WebSocketServerCallbacks> callbacks_;
  std::unique_ptr<FetchAssetHandler> fetch_asset_;
  std::unique_ptr<foxglove_websocket_server, foxglove_error (*)(foxglove_websocket_server*)> impl_;
};

}  // namespace foxglove
