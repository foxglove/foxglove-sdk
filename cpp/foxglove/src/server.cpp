#include <foxglove-c/foxglove-c.h>
#include <foxglove/server.hpp>

namespace foxglove {

WebSocketServer::WebSocketServer(WebSocketServerOptions options)
    : _impl(foxglove_server_start(options.name.c_str(), options.host.c_str(), options.port),
            foxglove_server_free) {}

void WebSocketServer::stop() {
  foxglove_server_stop(_impl.get());
}

}  // namespace foxglove
