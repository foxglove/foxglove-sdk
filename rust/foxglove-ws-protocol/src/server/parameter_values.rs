use serde::{Deserialize, Serialize};

use crate::parameter::Parameter;
use crate::JsonMessage;

/// Parameter values message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#parameter-values>
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "parameterValues", rename_all = "camelCase")]
pub struct ParameterValues {
    /// Parameter values.
    pub parameters: Vec<Parameter>,
    /// ID from a get/set parameters request.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

impl ParameterValues {
    /// Creates a new parameter values message.
    pub fn new(
        parameters: impl IntoIterator<Item = Parameter>,
        id: impl Into<Option<String>>,
    ) -> Self {
        Self {
            parameters: parameters.into_iter().collect(),
            id: id.into(),
        }
    }
}

impl JsonMessage for ParameterValues {}

#[cfg(test)]
mod tests {
    use crate::server::ServerMessage;

    use super::*;

    fn message() -> ParameterValues {
        ParameterValues {
            id: None,
            parameters: vec![
                Parameter::empty("empty"),
                Parameter::float64("f64", 1.23),
                Parameter::float64_array("f64[]", vec![1.23, 4.56]),
                Parameter::byte_array("byte[]", [0x10, 0x20, 0x30]),
                Parameter::bool("bool", true),
            ],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ParameterValues(orig));
    }
}
