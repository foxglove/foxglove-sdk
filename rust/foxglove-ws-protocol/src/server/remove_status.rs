use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// Remove status message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#remove-status>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "removeStatus", rename_all = "camelCase")]
pub struct RemoveStatus {
    /// IDs of the status messages to be removed.
    pub status_ids: Vec<String>,
}

impl JsonMessage for RemoveStatus {}

#[cfg(test)]
mod tests {
    use crate::server::ServerMessage;

    use super::*;

    fn message() -> RemoveStatus {
        RemoveStatus {
            status_ids: vec!["status-1".into(), "status-2".into(), "status-3".into()],
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::RemoveStatus(orig));
    }
}
