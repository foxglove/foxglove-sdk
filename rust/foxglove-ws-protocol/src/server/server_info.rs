//! Server info message types.

use std::collections::HashMap;

use crate::JsonMessage;
use serde::{Deserialize, Serialize};

/// A capability advertised in a [`ServerInfo`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Capability {
    /// Allow clients to advertise channels to send data messages to the server.
    ClientPublish,
    /// Allow clients to get & set parameters.
    Parameters,
    /// Allow clients to subscribe to parameter changes.
    ParametersSubscribe,
    /// The server may publish binary time messages.
    Time,
    /// Allow clients to call services.
    Services,
    /// Allow clients to subscribe to updates to the connection graph.
    ConnectionGraph,
    /// Allow clients to fetch assets.
    Assets,
}

/// Server info message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#server-info>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "serverInfo", rename_all = "camelCase")]
pub struct ServerInfo {
    /// Free-form information about the server.
    pub name: String,
    /// The optional features supported by this server.
    pub capabilities: Vec<Capability>,
    /// The encodings that may be used for client-side publishing or service call
    /// requests/responses. Only set if client publishing or services are supported.
    pub supported_encodings: Vec<String>,
    /// Optional map of key-value pairs.
    pub metadata: HashMap<String, String>,
    /// Optional string.
    pub session_id: String,
}

impl JsonMessage for ServerInfo {}

#[cfg(test)]
mod tests {
    use crate::server::ServerMessage;

    use super::*;

    fn message() -> ServerInfo {
        ServerInfo {
            name: "example server".into(),
            capabilities: vec![Capability::ClientPublish, Capability::Time],
            supported_encodings: vec!["json".into()],
            metadata: maplit::hashmap! {
                "key".into() => "value".into(),
            },
            session_id: "1675789422160".into(),
        }
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::ServerInfo(orig));
    }
}
