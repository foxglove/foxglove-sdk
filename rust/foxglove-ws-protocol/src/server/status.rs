//! Status message types.

use crate::JsonMessage;
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

/// Level indicator for a [`Status`] message.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize_repr, Deserialize_repr)]
#[repr(u8)]
#[allow(missing_docs)]
pub enum Level {
    Info = 0,
    Warning = 1,
    Error = 2,
}

/// Status message.
///
/// Spec: <https://github.com/foxglove/ws-protocol/blob/main/docs/spec.md#status>
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "op", rename = "status", rename_all = "camelCase")]
pub struct Status {
    /// Log level.
    pub level: Level,
    /// Message.
    pub message: String,
    /// Optional identifier for the status message.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

impl Status {
    /// Creates a new status message.
    pub fn new(level: Level, message: impl Into<String>) -> Self {
        Self {
            level,
            message: message.into(),
            id: None,
        }
    }

    /// Creates a new info-level message.
    pub fn info(message: impl Into<String>) -> Self {
        Self::new(Level::Info, message)
    }

    /// Creates a new warning-level message.
    pub fn warning(message: impl Into<String>) -> Self {
        Self::new(Level::Warning, message)
    }

    /// Creates a new error-level message.
    pub fn error(message: impl Into<String>) -> Self {
        Self::new(Level::Error, message)
    }

    /// Sets the status message ID, so that this status can be replaced or removed in the future.
    #[must_use]
    pub fn with_id(mut self, id: impl Into<String>) -> Self {
        self.id = Some(id.into());
        self
    }
}

impl JsonMessage for Status {}

#[cfg(test)]
mod tests {
    use crate::server::ServerMessage;

    use super::*;

    fn message() -> Status {
        Status::warning("Oh no").with_id("status-123")
    }

    #[test]
    fn test_encode() {
        insta::assert_json_snapshot!(message());
    }

    #[test]
    fn test_roundtrip() {
        let orig = message();
        let buf = orig.to_string();
        let msg = ServerMessage::parse_json(&buf).unwrap();
        assert_eq!(msg, ServerMessage::Status(orig));
    }
}
