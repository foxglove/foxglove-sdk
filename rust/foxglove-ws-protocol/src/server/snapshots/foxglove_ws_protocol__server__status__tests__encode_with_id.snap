---
source: rust/foxglove-ws-protocol/src/server/status.rs
expression: message_with_id()
---
{
  "op": "status",
  "level": 1,
  "message": "Oh no",
  "id": "my-id"
}
